app/notebook-fs.js

```diff
@@ -0,0 +1,122 @@
+                    resolve(false);
+                } else {
+                    reject(err);
+                }
+            } else {
+                resolve(stats.isFile());
+            }
+        });
+    });
+}
+
+function _statError(err, nbpath) {
+    var e = new Error('Error getting notebook info: ' + nbpath + ' - ' + err.message);
+    if (err.code === 'ENOENT') {
+        e.status = 404;
+    }
+    return e;
+}
+
+function _statNbOrDir(nbpath) {
+    return new Promise(function(resolve, reject) {
+        // try to stat a notebook first, then fallback to original value
+        var extPath = appendExt(nbpath, DB_EXT);
+        fs.stat(extPath, function(err, stats) {
+            if (err && extPath === nbpath) {
+                reject(_statError(err, nbpath));
+            } else if (stats && stats.isFile()) {
+                stats.fullpath = extPath;
+                resolve(stats); // notebook exists
+            } else {
+                // path may be a directory (bundled dashboard) or other file
+                fs.stat(nbpath, function(err, stats) {
+                    if (err) {
+                        reject(_statError(err, nbpath));
+                    } else {
+                        stats.fullpath = nbpath;
+                        resolve(stats);
+                    }
+                });
+            }
+        });
+    });
+}
+
+function _stat(nbpath) {
+    nbpath = _ensureDataPath(nbpath);
+    var ext = path.extname(nbpath);
+
+    return _statNbOrDir(nbpath).then(function(stats) {
+        if (stats.isDirectory()) {
+            // check if this directory contains an index dashboard
+            var indexPath = path.join(nbpath, INDEX_NB_NAME);
+            return _exists(indexPath).then(function (exists) {
+                stats.isDashboard = stats.hasIndex = exists;
+                if (stats.hasIndex) {
+                    stats.fullpath = indexPath;
+                    return new Promise(function(resolve) {
+                        // check if bundled dashboard has an 'urth_components' dir
+                        fs.stat(path.join(nbpath, 'urth_components'),
+                            function(err, urth_stats) {
+                                stats.supportsDeclWidgets = !err && urth_stats.isDirectory();
+                                resolve(stats);
+                            }
+                        );
+                    });
+                } else {
+                    return stats;
+                }
+            });
+        } else {
+            // might be a dashboard file or a "regular" file
+            stats.isDashboard = stats.isFile() && path.extname(stats.fullpath) === DB_EXT;
+            return stats;
+        }
+    });
+}
+
+module.exports = {
+    /**
+     * Checks if the specified file exists
+     * @param  {String} nbpath - path to a notebook
+     * @return {Promise(Boolean)} resolved to true if it exists
```

Comment:
It looks like they have defined [a syntax](https://github.com/jsdoc3/jsdoc/issues/1110) for documenting Promises but not implemented yet. In this example it would be `{Promise.<boolean>}`. I can update all the doc strings to this syntax.